search/streets_matcher.cpp

```diff
@@ -0,0 +1,165 @@
+#include "search/streets_matcher.hpp"
+#include "search/features_filter.hpp"
+#include "search/house_numbers_matcher.hpp"
+#include "search/query_params.hpp"
+
+#include "indexer/search_string_utils.hpp"
+
+#include "base/logging.hpp"
+#include "base/stl_helpers.hpp"
+
+namespace search
+{
+namespace
+{
+bool LessByHash(StreetsMatcher::Prediction const & lhs, StreetsMatcher::Prediction const & rhs)
+{
+  if (lhs.m_hash != rhs.m_hash)
+    return lhs.m_hash < rhs.m_hash;
+
+  if (lhs.m_prob != rhs.m_prob)
+    return lhs.m_prob > rhs.m_prob;
+
+  if (lhs.GetNumTokens() != rhs.GetNumTokens())
+    return lhs.GetNumTokens() > rhs.GetNumTokens();
+
+  return lhs.m_startToken < rhs.m_startToken;
+}
+}  // namespace
+
+// static
+void StreetsMatcher::Go(BaseContext const & ctx, FeaturesFilter const & filter,
+                        QueryParams const & params, vector<Prediction> & predictions)
+{
+  size_t const kMaxNumPredictions = 3;
+  double const kTailProbability = 0.05;
+
+  predictions.clear();
+  FindStreets(ctx, filter, params, predictions);
+
+  if (predictions.empty())
+    return;
+
+  sort(predictions.begin(), predictions.end(), &LessByHash);
+  predictions.erase(
+      unique(predictions.begin(), predictions.end(), my::EqualsBy(&Prediction::m_hash)),
+      predictions.end());
+
+  sort(predictions.rbegin(), predictions.rend(), my::LessBy(&Prediction::m_prob));
+  while (predictions.size() > kMaxNumPredictions && predictions.back().m_prob < kTailProbability)
+    predictions.pop_back();
+}
+
+// static
+void StreetsMatcher::FindStreets(BaseContext const & ctx, FeaturesFilter const & filter,
+                                 QueryParams const & params, vector<Prediction> & predictions)
+{
+  for (size_t startToken = 0; startToken < ctx.m_numTokens; ++startToken)
+  {
+    if (ctx.m_usedTokens[startToken])
+      continue;
+
+    // Here we try to match as many tokens as possible while
+    // intersection is a non-empty bit vector of streets. Single
+    // tokens that are synonyms to streets are ignored.  Moreover,
+    // each time a token that looks like a beginning of a house number
+    // is met, we try to use current intersection of tokens as a
+    // street layer and try to match BUILDINGs or POIs.
+    CBV streets(ctx.m_streets);
+
+    CBV all;
+    all.SetFull();
+
+    size_t curToken = startToken;
+
+    // This variable is used for prevention of duplicate calls to
+    // CreateStreetsLayerAndMatchLowerLayers() with the same
+    // arguments.
+    size_t lastToken = startToken;
+
+    // When true, no bit vectors were intersected with |streets| at all.
+    bool emptyIntersection = true;
+
+    // When true, |streets| is in the incomplete state and can't be
+    // used for creation of street layers.
+    bool incomplete = false;
+
+    auto emit = [&]()
```

Comment:
I guess the main rule that beats all that "X is a bad practice..." is "be reasonable." I agree - it's bad to declare a lambda that captures everything and then to pass it to external API. But please, take a closer look at how emit is used - it's used locally, to prevent code duplication, as it's body is complex and it's called multiple times. With the same effect I can declare a macro and then call it several times.

To sum up, I see no benefit in explicit capture specification.